runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/control/ProcessBundleDescriptors.java

```diff
@@ -141,9 +139,7 @@ private static ExecutableProcessBundleDescriptor fromExecutableStageInternal(
 
     Map<String, Map<String, TimerSpec>> timerSpecs = forTimerSpecs(stage, components);
 
-    if (bagUserStateSpecs.size() > 0 || timerSpecs.size() > 0) {
-      lengthPrefixKeyCoder(stage.getInputPCollection().getId(), components);
-    }
+    lengthPrefixAnyInputCoder(stage.getInputPCollection().getId(), components);
```

Comment:
Worth mentioning this will break any existing non-stateful pipelines. We could add another exception here for SDFs, but it's probably better to have a consistent encoding which does not depend on DoFns using state or timers.

Reply:
I would not expect it will break non-stateful pipelines since non-stateful transform doesn't look into this coder?

Reply:
It does, e.g. for buffering data during checkpoints. However, there are usually also other changes in Beam which makes restoring state from previous Beam versions hard.

Reply:
In that case, I would image that the buffering data should be encoded/decoded with grpc input coder, which has been lengthperfixed already. Would you like to elaborate more on this?

Reply:
Or do we have some test cases that we can verify?

Reply:
My bad, the length prefixing here only applies to the coder in the ProcessBundleDescriptor. Indeed, any buffering won't be affected by this because the elements have already been decoded when they are buffered.

For the test case, FlinkSavepointITCase should verify that.